### PR TITLE
Feature additions

### PR DESCRIPTION
**Pull request by eywalker**
* Fixed bug in `SpatialXFeatureLinear` where the normalization and constraining of spatial weights were not properly applied
* Added extension to `torch.nn.Module` where modules will now have `constrain_all` method that can be used to trigger `constrain` call on all submodules. This is WIP and the API *will change*.
* Added `ExtendedConv2d` with `SAME` padding option as well as `ConstrainedConv2d` and `ConstrainedConvTranspose2d` where constraints can be applied to the weight variable everytime `forward` is called
* Added `get_conv` utility function that would return properly configured convolution module that will satisfy the specified input size, output size, and kernel size (and strides if specified as well). 
* `conv2d_config` is the underlying utility function that figures out the necessary type of convolution (normal vs transpose), stride size, input and output paddings to achieve the specified input size, output size, and kernel size.
   * API for `conv2d_config` is a subject to change and ought not to be relied on. In contrast, the `get_conv` API is expected to be relatively stable with only additional arguments added in the future.

### CODE REVIEW

File: attorch/constraints.py [hunk 1]
```diff
@@ -1,16 +1,33 @@
...
+
 
-def negative(weight):
+def negative(weight, cache=None):
     weight.data *= weight.data.le(0).float()
+    return cache
 
-def positive_except_self(weight):
+
+def positive_except_self(weight, cache=None):
```
Comment (fabiansinz) [inline attorch/constraints.py]: Is it on purpose that `cache` is not used in all three constraints?

File: attorch/layers.py [hunk 2]
```diff
@@ -212,3 +213,160 @@ def __init__(self, features, **kwargs):
...
+    """
+    Extended 2D convolution module with fancier padding options.
+    """
+
+    def __init__(self, in_channels, out_channels, kernel_size, stride=1,
+                 padding=0, in_shape=None, groups=1, bias=True):
+        kernel_size = _pair(kernel_size)
+        stride = _pair(stride)
+
+        if padding == 'SAME':
```
Comment (fabiansinz) [inline attorch/layers.py]: I propose to make it mode numpy style and use lower case strings.
